### PR TITLE
Specification for logged data

### PR DESCRIPTION
**Pull request by joaander**
## Description

* Add user-defined logged data section to the `hoomd` specification
* Add API to find chunk names that are present in a gsd file

## Motivation and Context

Many users (in offline discussions) have requested the ability to log custom data to gsd files. In most cases, it is better to compute this data when needed - but in some situations it is necessary to save data computed at simulation time. This PR proposes a naming scheme for custom logged data in GSD files. HOOMD will later implement a python callback mechanism for feeding user-defined data into user-defined chunk names under `log/`.

With the added API to discover chunk names, visualization and analysis can discover what quantities users have logged and make them available. To aid these tools, the specification suggests using the prefix `log/particles/` for per-particle quantities.

## How Has This Been Tested?

* Added unit tests.
* Documentation preview: 
  * https://gsd.readthedocs.io/en/log-data/schema-hoomd.html#logged-data
  * https://gsd.readthedocs.io/en/log-data/c-api.html#c.gsd_find_matching_chunk_name
  * https://gsd.readthedocs.io/en/log-data/fl-examples.html#discover-chunk-names
  * https://gsd.readthedocs.io/en/log-data/hoomd-examples.html#access-logged-data

## Change log

```
* File API: Add ``find_matching_chunk_names()``
*  ``HOOMD`` schema 1.4: Add user defined logged data
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/gsd/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**GSD Contributor Agreement**](https://github.com/glotzerlab/gsd/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/gsd/blob/master/doc/credits.rst).

## Note

This PR is based on #37. I will rebase on master after #37 is merged.

### CODE REVIEW

File: doc/hoomd-examples.rst [hunk 1]
```diff
@@ -141,3 +141,31 @@ format. In this example, type 'A' has 3 vertices (the first 3 in the list) and t
...
+
+Access logged data
+^^^^^^^^^^^^^^^^^^
+
+.. ipython:: python
+
+    with gsd.hoomd.open(name='example.gsd', mode='wb') as t:
+        s = gsd.hoomd.Snapshot()
+        s.particles.N = 4
+        s.log['particles/net_force'] = [[-1,2,-3],
```
Comment (bdice) [inline doc/hoomd-examples.rst]: I think this data is being saved as an array of integers, based on the formatting -- this choice of an example array might be a little confusing to a user who is expecting to read/write float values for forces. I would change the values to include decimals or create a NumPy array with the appropriate type.

File: doc/schema-hoomd.rst [hunk 2]
```diff
@@ -12,12 +12,12 @@ of the existing data chunks. Any newer reader will initialize new data chunks wi
...
 
 :Schema name: ``hoomd``
-:Schema version: 1.3
+:Schema version: 1.4
 
 Use-cases
 ---------
 
-There are a few problems with XML, DCD, and other dump files that the GSD schema ``hoomd`` solves.
+The GSD schema ``hoomd`` provides.
```
Comment (bdice) [inline doc/schema-hoomd.rst]:
```suggestion
The GSD schema ``hoomd`` provides:
```

Comment (bdice) [reply doc/schema-hoomd.rst]: The contents of the list below don't finish the sentence started above with "provides."

File: tests/test_hoomd.py [hunk 3]
```diff
@@ -496,3 +496,33 @@ def test_state(tmp_path):
 
         numpy.testing.assert_array_equal(s.state['hpmc/convex_polyhedron/N'], snap1.state['hpmc/convex_polyhedron/N']);
         numpy.testing.assert_array_equal(s.state['hpmc/convex_polyhedron/vertices'], snap1.state['hpmc/convex_polyhedron/vertices']);
+
+def test_log(tmp_path):
+    snap0 = gsd.hoomd.Snapshot();
+
+    snap0.log['particles/net_force'] = [[1,2,3],[4,5,6]]
+    snap0.log['value/potential_energy'] = [10]
```
Comment (lyrivera) [inline tests/test_hoomd.py]: Following @bdice's previous comments, perhaps we can use NumPy arrays here too. Same applies below.

Comment (joaander) [reply tests/test_hoomd.py]: Feel free to make the change if you want. The specification assigns no particular data types for any particular chunk names. This unit test correctly validates that data, as defined by implicit numpy conversion rules, survives the round trip from write to read.

**Comment by joaander** [comment]
A stakeholder requested that logged data provide default values from frame 0 to save on file space for static logged quantities. Is the new wording in `schema-hoomd.rst` clear in 
6e9eeb5 ? If not, please propose or commit an alternative.

**Comment by bdice** [comment]
> A stakeholder requested that logged data provide default values from frame 0 to save on file space for static logged quantities. Is the new wording in `schema-hoomd.rst` clear in
> [6e9eeb5](https://github.com/glotzerlab/gsd/commit/6e9eeb5c8ddacbd817a13709499fbb2c3966290b) ? If not, please propose or commit an alternative.

Just to clarify my understanding: before that commit, it was valid to only log a quantity on frame 1 (but not frame 0), and after that commit, logging only on frame 1 (but not frame 0) gives an invalid GSD file?

I think that wording is fine, if my description above is correct.

**Comment by joaander** [comment]
> I think that wording is fine, if my description above is correct.

Correct. Frame 0 defines the default in the new wording.

**Comment by lyrivera** [comment]
> A stakeholder requested that logged data provide default values from frame 0 to save on file space for static logged quantities. Is the new wording in `schema-hoomd.rst` clear in
> [6e9eeb5](https://github.com/glotzerlab/gsd/commit/6e9eeb5c8ddacbd817a13709499fbb2c3966290b) ? If not, please propose or commit an alternative.

I think the wording is clear

**Comment by joaander** [comment]
Ready for final review.

File: tests/test_hoomd.py (hunk 3 above)
Comment (lyrivera) [reply tests/test_hoomd.py]: In that case we can leave it as is.

**Pull request merged**